Add play/pause audio in notes

Abanoub-Asaad opened the pull request:
## Purpose / Description
`Before` The learner can only replay the audio in notes.
`After` The learner can pause or resume the audio and It works with any number of audio files in the note.

## Fixes
Fixes #6774 

## Approach
Focusing on the URL of the audio. When the user clicks on another icon ⏯️, I'll work on the current URL not the previous one.

## How Has This Been Tested?
Using a real mobile phone "Xiaomi".

## Demo video

https://user-images.githubusercontent.com/48657780/117230198-d7563300-ae1c-11eb-8ea3-2ea53a007b16.mp4

## Aditional Notes
Related PRs  #8346


## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

Akshay0701 commented:
This is great! 
it would be great if the play and pause icon is added

Abanoub-Asaad commented:
> This is good!
> can we add play and pause icons?

Check the "additional notes" section in the PR template

Akshay0701 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3702,31 @@ private boolean filterUrl(String url) {
 
 
+        private void controlSound(String url) {
+            //check if I click on another audio icon instead of the current icon.
```
```suggestion
            // check if I click on another audio icon instead of the current icon.
```

TarekkMA commented:
[Opinon] Extract the sound handling logic to a new class (maybe `SoundController`) under this package `com.ichi2.anki.cardviewer` under new class, so it's unit testable

Abanoub-Asaad commented:
> [Opinon] Extract the sound handling logic to a new class (maybe `SoundController`) under this package `com.ichi2.anki.cardviewer` under new class, so it's unit testable

Good idea! Thank you, Tarek.
Also, I want feedback from @mikehardy  and @david-allison-1.

i-zanis commented:
This is all looking brilliant but does this work with the TTS?

It beats the purpose of creating it if it doesn't.

Abanoub-Asaad commented:
> This is all looking brilliant but does this work with the TTS?
> The icons are not needed. The user can find the function in the menu.
> He can find these functions to hand gestures as he can now with the replay button.
> 
> It beats the purpose of creating it if it doesn't.

It's not in the Hand Gestures now, but only in the default status like when you're trying to add, preview or study an audible note. But I'll add it into hand gestures when this PR got reviewed and merged.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3702,31 @@ private boolean filterUrl(String url) {
+                mSoundPlayer.pauseSound();
+            } else if (!mFlagCurrentAudio) {
+                mFlagCurrentAudio = !mFlagCurrentAudio;
+                mIsMediaPlaying = !mIsMediaPlaying;
```
Since we now that `mFlagCurrentAudio` is False, then using `mFlagCurrentAudiod=True` would seem more clear to me.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3702,31 @@ private boolean filterUrl(String url) {
+            }
+
+            if (mIsMediaPlaying) {
+                mIsMediaPlaying = !mIsMediaPlaying;
```
Since `mIsMediaPlaying = !mIsMediaPlaying;` occur in each branch, I think you can just add it at the end of the method and avoid duplicate code

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -317,6 +317,12 @@
 
     private boolean mButtonHeightSet = false;
 
+    // To check whether media is playing or not
```
Usually, the comment should indicate the value of the variable, not the reason we use it. If it's really hard to follow, then the comment should be at the class level to help understand the way the whole logic work.

The variable name `mIsMediaPlaying` is already quite clear. So no need to comment it.

Arthur-Milchior replied:
Did you try to play a sound entirely and then reclick on the button? What does it do in this case?

I don't see anything that ensure that at the end of the play, `mIsMediaPlaying` moves to false. Did I miss something?

I think you should uses https://developer.android.com/reference/android/media/MediaPlayer#isPlaying() instead (you'll nead to add a method in Sound.java that allow to call this method on the MediaPlayer)

Abanoub-Asaad replied:
Yes, I missed this part
But I'm going to solve it.

Arthur-Milchior replied:
Thanks

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3516,10 +3514,7 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
         // We play sounds through these links when a user taps the sound icon.
         private boolean filterUrl(String url) {
             if (url.startsWith("playsound:")) {
-                // Send a message that will be handled on the UI thread.
```
I think it would be nice to extract those three lines into a method in a first commit. Then you can call this method from `controlSound` in a second commit.

Arthur-Milchior replied:
I fear I was misconstrued. It's not that important however, just an opinion to improve the quality. My idea was to first move this code, so that it would a very simple commit to review, then to create `controlSound`.
By creating `controlSound` first, you make the commit "Add play/pause audio in notes" harder than needed

Abanoub-Asaad replied:
In the current code, I'm using those three lines twice, so I think it's better to be in a separate method.

Arthur-Milchior replied:
Please, either resolve or answer but don't do both. Otherwise I don't see your answer.

I agree that it may be nice in a method. I was just making a comment about the order of your commits. Making this function should be done first

Abanoub-Asaad replied:
Sorry for that and thank you for letting me know.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -317,6 +317,12 @@
 
+    // To check whether media is playing or not
+    private boolean mIsMediaPlaying = false;
+    // To check if the user is playing one audio and then playing another one or not
```
I don't understand why `mFlagCurrentAudio` need to be a member ? I would thought it would suffice to define it locally since it's used in a single function. 
If you could describe the value of the variable instead of its goal, it would be nice

Abanoub-Asaad replied:
I have deleted `mFlagCurrentAudio` and minimized the code size.

Arthur-Milchior replied:
THanks

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -286,6 +286,24 @@ public void playSound(String soundPath, OnCompletionListener playAllListener, fi
+    /**
+     * Pauses the given sound. Called on pressing the content inside span tag.
+     */
+    public void pauseSound() {
```
I'd create an extra function `playOrPause` that check whether the media is playing, if so I'd pause, otherwise I'd play. This ensure that the same function can be called on each click on your button. Then your button would only have to check if the media to play now is the same than last time.

Arthur-Milchior reviewed (commented):
First. Thank you very very much. That will be extremely helpful!
I use anki to practice music. I tried to ask it to have a sound and then play it, and having to deal with volume is clearly less great than being able to pause and restart (actually, I'd love to restart from start, and that may be a great option to add, but that's for another pr if we ever do it)


When a music end, does the next start automatically on first run? This is what current version does, and I'd like to check whether it is still the case.

Also, did you try waiting for last music to finish, and then click to replay the last music?

> If I added another icon "pause icon", I would face some bugs. That's why I haven't added more icons
Also, the Anki Desktop doesn't include another icon when pausing the audio.

Can you please let me know what bug you encountered? Maybe we can help

Abanoub-Asaad commented:
> First. Thank you very very much. That will be extremely helpful!
> I use anki to practice music. I tried to ask it to have a sound and then play it, and having to deal with volume is clearly less great than being able to pause and restart (actually, I'd love to restart from start, and that may be a great option to add, but that's for another pr if we ever do it)
> 
> When a music end, does the next start automatically on first run? This is what current version does, and I'd like to check whether it is still the case.
> 
> Also, did you try waiting for last music to finish, and then click to replay the last music?
> 
> > If I added another icon "pause icon", I would face some bugs. That's why I haven't added more icons
> > Also, the Anki Desktop doesn't include another icon when pausing the audio.
> 
> Can you please let me know what bug you encountered? Maybe we can help

Thank you for your feedback.

After what I've done, When the current music ends, nothing will start automatically.

After waiting for the current song's end, I couldn't replay the song again, **but I'll work to solve this bug.**

The problem when I was adding icons for pausing and resuming audio is :
The only icon changes is the first left one, but nothing else
``for instance``, If I have 3 icons and click on the first left icon, it will change successfully 
But when I click on the second or third icon, Yes the audio pause and resume successfully but the icons don't change and the first icon changes after each click on the other icons.

Arthur-Milchior commented:
Do you know javascript and html?
It seems a standard javascript issue here. 

You should add an id to each span, so instead of looking for the first span, you can look for specific id.
And also, you should create a js function that takes the number of the audio to play, and change the pause/play button. This way, you can call the function from java instead of having to send a big bunch of code. The main advantage being that javascript code will be formatted as javascript and easy to read.

Abanoub-Asaad commented:
> Do you know javascript and html?
> It seems a standard javascript issue here.
> 
> You should add an id to each span, so instead of looking for the first span, you can look for specific id.
> And also, you should create a js function that takes the number of the audio to play, and change the pause/play button. This way, you can call the function from java instead of having to send a big bunch of code. The main advantage being that javascript code will be formatted as javascript and easy to read.

I know HTML and a little Javascript, I'll check this out.
What do you think about the commits that I've done after your comments?

Abanoub-Asaad commented:
Could you review it now, @Arthur-Milchior ?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3710,12 +3710,12 @@ private void controlSound(@Nullable String url) {
             }
 
             if (mIsMediaPlaying) {
```
`mIsMediaPlaying || mFlagCurrentAudio` 

This way, the last else become unneeded

Arthur-Milchior reviewed (commented):
I am really sorry, but it begins to be really hard to review, because of the number of commits.
You add things and then remove then.

It would be nice if you could use to corrects commit instead of adding new one, so that the pull request contain the minimum number of changes and all changes are meaningful. Right now, I don't know in the last commit what is there only to correct remark I made yesterday and what is here because that introduce new feature or clean code (e.g. ensuring that after a sound, the next one plays)


I would have a preference for the commit that change the image to be in the same PR, because I find the current interface a little bit confusing. The pause button is really a well established standard and not having it make little sens. But we can discuss this when this PR is improved already

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3699,33 @@ private boolean filterUrl(String url) {
+                onCurrentAudioChanged(url);
+            } else if (mIsMediaPlaying) {
+                mSoundPlayer.playOrPauseSound();
+            } else {
```
This last if/else is useless since both branches do the same thing

Arthur-Milchior replied:
I'm sorry, but I don't think it's really resolved.

You can but `mCurrentUrl = url` at the end of the method since it will always be correct.
Then you can do:
```java
 if (!url.equals(mCurrentUrl) || mSoundPlayer.mCurrentAudioIsFinished) {
￼                onCurrentAudioChanged(url);
} else {
    mSoundPlayer.playOrPauseSound();
}
mCurrentUrl = url;
```

This lead to shorter code, no code repetition, and exactly the same behavior

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -317,6 +317,9 @@
 
     private boolean mButtonHeightSet = false;
 
+    private boolean mIsMediaPlaying = false;
```
I beg your pardon, I thought I wrote a comment and it seems I can't find it anymore 
I don't see the point of `mIsMediaPlaying`. The Media player already has a method to determines if anything is playing right now. It can be used instead.

Actually, there is a problem here. When the sound ends, `mIsMediaPlaying` is not set back to false (or I missed it). So if you wait for a sound to ends and press to play it again, it won't be detected as a request to play a second time

Abanoub-Asaad commented:
> I am really sorry, but it begins to be really hard to review, because of the number of commits.
> You add things and then remove then.
> 
> It would be nice if you could use to corrects commit instead of adding new one, so that the pull request contain the minimum number of changes and all changes are meaningful. Right now, I don't know in the last commit what is there only to correct remark I made yesterday and what is here because that introduce new feature or clean code (e.g. ensuring that after a sound, the next one plays)
> 
> I would have a preference for the commit that change the image to be in the same PR, because I find the current interface a little bit confusing. The pause button is really a well established standard and not having it make little sens. But we can discuss this when this PR is improved already

I'm sorry for not organizing the commits, but I have organized and refactored everything in one commit now.

I've fixed this issue: when the audio finishes, you can replay it again.

I haven't found this feature before in AnkiDroid: when the audio finishes, the next one will play automatically. If I'm wrong, Can you please correct me?

Arthur-Milchior commented:
https://user-images.githubusercontent.com/357361/118141994-40840a80-b40a-11eb-8e05-5f5ac839cdab.mp4

This card play one media after the other. This as always been the default behavior on my phone

Arthur-Milchior commented:
So, i’d love to get a video with two sound. Opening the card and letting both sound play entirely, then clicking on your button, letting the sound endss, then reclicking on the same button, pausing, restarting, pausin, playing the second sound, replaying the first one

This should ensure most features are actually tested please

Abanoub-Asaad commented:
>  Screen_video_2021-05-13-16-41-10.mp4 
> This card play one media after the other. This as always been the default behavior on my phone

I'm really sorry, the whole problem happened because of My mobile "Xiaomi"
When I have used another mobile everything works fine as you said.

Abanoub-Asaad commented:
> So, i’d love to get a video with two sound. Opening the card and letting both sound play entirely, then clicking on your button, letting the sound endss, then reclicking on the same button, pausing, restarting, pausin, playing the second sound, replaying the first one
> 
> This should ensure most features are actually tested please


https://user-images.githubusercontent.com/48657780/118201833-b44d0400-b458-11eb-980c-d7257cc14df5.mp4

Could you tell me your feedback, @Arthur-Milchior ?

Abanoub-Asaad commented:
Could you tell me your feedback, @Arthur-Milchior?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -72,6 +72,9 @@
      */
     private MediaPlayer mMediaPlayer;
 
+    public boolean mCurrentAudioIsFinished;
```
In Java, it is very exceptional for a variable to be public and not final. Usually, you make the variable private and add a getter to it. This ensure that nobody, in the future, from another class, change the variable's value.

I think it would be easier to create a method
```java
public boolean isAudioFinished() {
   return mMediaPlayer == null;
}
```
This would avoid to add an extra variable. It's usually good to avoid adding extra state unless it's costly to recompute.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3698,29 @@ private boolean filterUrl(String url) {
+        }
+
+
+        private void onCurrentAudioChanged(@Nullable String url) {
```
I'm going to insist, but I think you could obtain more readable and simpler commit by creating this method in a first commit (and calling it from `filterUrl`) before creating this big commit.

Abanoub-Asaad replied:
Sorry, I couldn't understand.
What do you mean by "commit", is it a `GitHub commit`?
And, Why do you want me to call it from `filterUrl`, and I'm calling it from `controlSound` ?

Arthur-Milchior replied:
It is a git commit. It is shown in github too, but that's a general git feature.

If you look at the history of the app commit list, it should read like a story explaining what was changed, ideally each explanation is as simple as possible.

So, what I am asking here is, intuitively:
> First, I extracted code from `filterUrl`; there is a part that is self consistent which deal with a change in audio. It is moved to a new method `onCurrentAudioChanged`
> Then, I added an intermediate step between `filterUrl` and `onCurrentAudioChanged`: the method `controlSound`, that ensure that `onCurrentAudioChanged` is only called if the audio is actually changed. Otherwise it calls `playOrPauseSound`, which is also a new method.

This seems easier to read to me than having all done simultaneously.

Actually, you can also add `playOrPauseSound` in a commit number 2 and the remaining of the changes in commit number 3

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2390,6 +2383,7 @@ protected void playSounds(boolean doAudioReplay) {
 
     private void playSounds(SoundSide questionAndAnswer) {
         mSoundPlayer.playSounds(questionAndAnswer, getSoundErrorListener());
+        mCurrentUrl = "playsound:" + mSoundPlayer.mFirstAudioUri;
```
Sorry, but I don't understand why the url should be the url of the first sound played. 
I would imagine that when you play a second sound automatically, the URL should switch to the second sound played and I don't see it occurring in the code.

I suspect it will create a problem in the following case.

Create a card with sounds A and B.

Let A ends. When B is playing, press A. In this case the sound will pause (because it will be equal to mCurrentUrl).

Close the reviewer and open it again to restart from scratch. Let A ends. When B is playing, press B. It will play B from start, because it has not detected that B is actually playing.

Abanoub-Asaad replied:
Actually, I've detected this problem and thought that I commented about it after uploading the previous video.
But I'm really sorry about that.

Arthur-Milchior reviewed (commented):
thanks a lot for the video. It seems to really work well, that's cool. Really useful feature ! 

I'm happy you found out why we had different experience. Really now idea why a specific brand behave differently, but as you may have heard by now, it's something we had to be used to. 

I would like you to do a test, please, with two sounds A and B. Please choose long sounds.

Create two cards C1 and C2 with sounds A and B in this order. When everything is played, click on B. While B is playing, review the card C1 so that C2 is shown. Then, when A is still playing on card C2, click on B.

I am not sure, but I suspect that when you press on B, the sound will just stop playing entirely. If you press B again, I suspect A will resume. I'm not entirely sure.
The idea is that, by doing so, you would have `mCurrentUrl` still being the url of B, even if it is A which is playing. So it will not detect it should switch from A to B.


I added another case that I believe will fail. I have not actually tested and may be totally wrong. I would be pleased to know that I'm wrong, and in this case I'll need to study more the logic of your code. Can you please do all the tests and let me know the result (ideally with a video)

Abanoub-Asaad commented:
> thanks a lot for the video. It seems to really work well, that's cool. Really useful feature !
> 
> I'm happy you found out why we had different experience. Really now idea why a specific brand behave differently, but as you may have heard by now, it's something we had to be used to.
> 
> I would like you to do a test, please, with two sounds A and B. Please choose long sounds.
> 
> Create two cards C1 and C2 with sounds A and B in this order. When everything is played, click on B. While B is playing, review the card C1 so that C2 is shown. Then, when A is still playing on card C2, click on B.
> 
> I am not sure, but I suspect that when you press on B, the sound will just stop playing entirely. If you press B again, I suspect A will resume. I'm not entirely sure.
> The idea is that, by doing so, you would have `mCurrentUrl` still being the url of B, even if it is A which is playing. So it will not detect it should switch from A to B.
> 
> I added another case that I believe will fail. I have not actually tested and may be totally wrong. I would be pleased to know that I'm wrong, and in this case I'll need to study more the logic of your code. Can you please do all the tests and let me know the result (ideally with a video)

Excuse me. I've a small question, Do you mean by "Creating 2 cards C1 and C2" something like that in the attached image?
I think it's a note that includes the question "C1" and the answer "C2", Am I right? 
<div align="center">
<pre>
   
<img alt="" width="30%" src="https://user-images.githubusercontent.com/48657780/118370937-57f9f980-b5aa-11eb-8640-89eb5a0ee9af.jpeg"> 
     
</pre>
</div>

Abanoub-Asaad commented:
Can you check this to record the video, @Arthur-Milchior ?

Arthur-Milchior commented:
I actually do not care if both cards comes from the same note or distinct note. That is why I did not mention it. But I agree it was not clear and it was credible that I meant either "two cards of a note" or "two notes".

Abanoub-Asaad commented:
https://user-images.githubusercontent.com/48657780/118492045-e8a61600-b71f-11eb-984e-047e78ab9a91.mp4

There's no problem because every time the user chooses a file, The file will have a different URL
But the problem would happen if the user copied the URLs.

Your thoughts @Arthur-Milchior ?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3707,6 +3698,25 @@ private boolean filterUrl(String url) {
         }
 
 
+        private void controlSound(@Nullable String url) {
```
Replace "I" by "the user" please

Arthur-Milchior replied:
Add javadoc. `controlSound` is really not specific enough.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3516,10 +3509,8 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
-                Message msg = Message.obtain();
-                msg.obj = url.replaceFirst("playsound:", "");
-                mHandler.sendMessage(msg);
+                mCurrentUrl = "playsound:" + mSoundPlayer.getCurrentAudioUri();
```
What is the point of `mCurrentUrl`. It seems to me that you could always simply use `mSoundPlayer.getCurrentAudioUri()`

Abanoub-Asaad replied:
`mCurrentUrl` was important in most previous commits.  After the current changes, it's useless so I've deleted it.

I've divided the previous commit into 3 commits.

Arthur-Milchior reviewed (commented):
Thanks a lot for those change. I'll need to test more over the week-end.
I think I am starting to understand what is going on. But I'd like more answer to be sure I'm not making any false assumptions.

If I'm right and `mCurrentUrl` is useless, then in the hand, no new member will be needed in `AbstractFlashcardViewer` and I would really love that. I didn't know why it would need more state, and it would clearly be easier to think about it if it actually do not use any more state

Arthur-Milchior commented:
Also, as @TarekkMA mentionned, adding more test would be great, to ensure it works as we expect. I won't require it for this PR, given how horrible our sound class is. But really is something that should be resolved at some point

Abanoub-Asaad commented:
> Also, as @TarekkMA mentionned, adding more test would be great, to ensure it works as we expect. I won't require it for this PR, given how horrible our sound class is. But really is something that should be resolved at some point

I'll try

Abanoub-Asaad commented:
Should I work on the icons now? or the icons will be in another PR that I'll make?

Sorry, I'm just referring those icons to be in another PR because the Anki Desktop doesn't have different icons in case of pausing or resuming the audio, that's why I'm thinking about making it in another PR. Your thoughts @Arthur-Milchior?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3703,12 +3694,25 @@ private boolean filterUrl(String url) {
+         * @param url
+         */
+        private void controlSound(String url) {
+            if (!url.equals("playsound:" + mSoundPlayer.getCurrentAudioUri()) || mSoundPlayer.isCurrentAudioFinished()) {
```
`getCurrentAudioUri` returns a Uri, so this sum implicitly call `toString()`. However, the javadoc does not make any promise about `StringUri` nor about `toString` so it seems very fragile, any change to the library would not break the current javadoc and would break this function. Furthermore, if `StringUri` were to change by making any modification, the inequality may become false because of a badly formatted Uri.

I think `getCurrentAudioUri()` should directly return the String. This diminish the risk to get a hard to understand bug in years, if `Uri` change. This seems a simple change to made in the same commit actually

Arthur-Milchior reviewed (commented):
I really appreciate the evolution of the PR. 
It's a good rule of thumbs that you should always avoid adding members if you can, as each member add complexity, makes the program harder to reason about. I find this quite easier to understand and assert its truth than the first version of the PR. Thank you very much for accepting to take the time to rework the PR and improve it so much!

I still have a few remarks, but things are getting quite better already.
Please let me know if you ever disagree or if you do not understand the change I ask you to make. I may always be wrong, and I hope that I make clear why I am asking all of this extra work on top of a working implementation.


I want to test this week-end before I merge. If you don't have anything more to do, I'd like you to work on the icons.

You can create a new branch, based on this branch (by the way, you should always create one branch for each PR and avoid to work directly on the branch master/main. Because it makes almost impossible to work on two PR in parallel, and it makes harder to rebase on top of ankidroid/master in case of conflict)

Abanoub-Asaad commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3703,12 +3694,25 @@ private boolean filterUrl(String url) {
             Message msg = Message.obtain();
             msg.obj = url.replaceFirst("playsound:", "");
-            mHandler.sendMessage(msg);
+            mSoundPlayer.playSound((String) msg.obj, null, null, getSoundErrorListener());
```
Thank you for pointing these things out.

As you said `Message` is not important after removing the `Handler`, I've deleted it after testing.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -72,6 +72,8 @@
      */
```
Can you add in the javadoc that mMedialPlayer is Nullable, and that it is set only if a sound is playnig or paused, otherwise it is null. This became very relevant because of your PR

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3703,12 +3694,23 @@ private boolean filterUrl(String url) {
-            Message msg = Message.obtain();
-            msg.obj = url.replaceFirst("playsound:", "");
-            mHandler.sendMessage(msg);
+            mSoundPlayer.playSound(url, null, null, getSoundErrorListener());
```
This seems good.
This change is atomic, I think it's worth having it's own commit.
 But it raises the important question of why it was not already this simple previously. Do you have any answer idea? Any research you made about this matter could be in the atomic commit comment

Arthur-Milchior reviewed (commented):
LGTM

Abanoub-Asaad commented:
> LGTM

Thank you

I've made the fourth commit, Is it good now?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -3703,12 +3694,26 @@ private boolean filterUrl(String url) {
-            msg.obj = url.replaceFirst("playsound:", "");
-            mHandler.sendMessage(msg);
+            msg.obj = url;
+            mSoundPlayer.playSound((String) msg.obj, null, null, getSoundErrorListener());
```
In the third commit, there is still too many things occurring at once for the commits to be really atomic. I'm actually confused slightly.

You δ-reduced `mHandler.sendMessage` (in plain English, you replaced a function by its definition.).
But when I re-read, I realize that `mSoundPlayer.stopSounds();` did disappear and I've no idea why.

If the δ-reduction was in a single atomic commit, I would have catched it more easily. And it would allow the commit message to explain why it was necessary.

Furthermore, it's not related to `controlSound` so it seems to create no problem to put both in different commit

And the more important questi on here is: why is it okay not to `stopSound` anymore?

Abanoub-Asaad replied:
Sorry, It needs more improvement.

I've organized the commits to be more specific.  Does it need more improvements now @Arthur-Milchior ?

Arthur-Milchior reviewed (commented):
Yes, thank you a lot. It gets more and more easy to concentrate on small part and be sure of them.

You choose yourself a complex part of the code. we have known for quite some time that sounds are a knot of old code that is hard to work in. This is simplifying it already, and I like that. This class may be deleted and rewritten one day maybe, but simply removing `mHandler` is already an improvement. It was really useles and it seems nobody noticed previously

Abanoub-Asaad commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -396,15 +396,6 @@
     // LISTENERS
     // ----------------------------------------------------------------------------
 
-    private final Handler mHandler = new Handler() {
-
-        @Override
-        public void handleMessage(Message msg) {
-            mSoundPlayer.stopSounds();
-            mSoundPlayer.playSound((String) msg.obj, null, null, getSoundErrorListener());
-        }
-    };
-
```
**Before my changes :** 
When I was clicking on the audio icon it was stopping the current sound using `mSoundPlayer.stopSounds();` and then playing it again using `mSoundPlayer.playSound` and these methods were implemented in the handler.

**After my changes :** 
Actually, I'm stopping the sound. When I click on the audio icon, it calls `playOrPauseSound()`. So when the audio is playing, it'll stop and vise versa.

Abanoub-Asaad replied:
Does this answer your question @Arthur-Milchior?

Arthur-Milchior replied:
Yes, thanks.
Can you please put it in the commit message (if not already done)

Abanoub-Asaad replied:
I've just updated the commit message.

Abanoub-Asaad commented:
@mikehardy, I'm excited to hear your feedback.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -286,11 +291,29 @@ public void playSound(String soundPath, OnCompletionListener playAllListener, fi
+     * Play or Pause the running sound. Called on pressing the content inside span tag.
+     */
+    public void playOrPauseSound() {
+        if (mMediaPlayer != null) {
```
two things: invert the if for readability
Get a reference to mMediaPlayer to ensure that it's not converted to null in another thread after the `if`

Abanoub-Asaad replied:
- Regarding the first point, I want to ensure that it's good now?
```java
public void playOrPauseSound() {
        if (mMediaPlayer == null) {
            return;
        }
        if (mMediaPlayer.isPlaying()) {
            mMediaPlayer.pause();
        } else {
            mMediaPlayer.start();
        }
    }
```
- Sorry! I didn't understand where to implement the second point.

david-allison-1 replied:
Second point: 

```java
MediaPlayer mediaPlayer = mMediaPlayer
if (mediaPlayer.isPlaying()) {
````

david-allison-1 reviewed (changes requested):
I'd love to see unit tests here.

Abanoub-Asaad commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -68,10 +68,15 @@
+    /**
+     * It's used to store the Uri of the current Audio in case of running or pausing.
+     */
+    private Uri mCurrentAudioUri = null;
```
I think no, because every time a card is flipped/loaded, `mCurrentAudioUri` will be set as the current audio path
This happened because every time a card is loaded, this method `playSoundInternal()` which is in the Sound class will be called.

david-allison-1 replied:
What happens when a card is loaded without sound?
Off the top of my head: load a card without sound, then "Replay Audio" gesture for example

Abanoub-Asaad replied:
If the card doesn't have a sound and I use the gestures to replay the audio, NOTHING will be playing as in this video
 
https://user-images.githubusercontent.com/48657780/119425679-1e429480-bd08-11eb-9923-ccf13ce62bac.mp4

Abanoub-Asaad commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -363,6 +386,10 @@ private void playSoundInternal(String soundPath, OnCompletionListener playAllLis
     }
 
+    public String getCurrentAudioUri() {
+        return mCurrentAudioUri.toString();
```
I think no
Because every time `mCurrentAudioUri` carries a Uri and converts it will be converted to a string.

Could you please tell me Why you think that it can throw an exception?

david-allison-1 replied:
> I think no
> Because every time `mCurrentAudioUri` carries a Uri and converts it will be converted to a string.
> 
> Could you please tell me Why you think that it can throw an exception?

Because `mCurrentAudioUri` can be null

Abanoub-Asaad commented:
@david-allison-1, What about now?

Abanoub-Asaad commented:
@david-allison-1, What about now?

Abanoub-Asaad commented:
Is this PR ready to get merged?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java:
```diff
@@ -286,11 +291,31 @@ public void playSound(String soundPath, OnCompletionListener playAllListener, fi
+        }
+    }
+
+    public boolean isCurrentAudioFinished() {
```
This could do with a comment inside the method explaining why it works

david-allison-1 reviewed (commented):
One question regarding "replay audio", then good to go

It seems fine, but confirmation would be fantastic.

Thanks!

david-allison-1 commented:
Just to note: I forgot to un-resolve: https://github.com/ankidroid/Anki-Android/pull/8781#discussion_r638353415

Should be a quick check and good to go

Abanoub-Asaad commented:
> Just to note: I forgot to un-resolve: [#8781 (comment)](https://github.com/ankidroid/Anki-Android/pull/8781#discussion_r638353415)
> 
> Should be a quick check and good to go

Ok, I've replied to the mentioned comment with a demo video.

david-allison-1 reviewed (approved):
LGTM, thanks for checking

david-allison-1 commented:
@Arthur-Milchior you added "second approval", but the review isn't checked, pinging as it's likely a very quick re-review

mikehardy commented:
Hi there!

Apologies for the delay, and if you're already submitted something to OpenCollective for May 2021 PRs merged, you may feel free to ignore this.

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for May 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in May)

Thanks!

WiliTest commented:
Is there a way to play/pause the TTS too?

mikehardy commented:
Android TTS APIs do not have pause/resume that I know of

WiliTest commented:
> Android TTS APIs do not have pause/resume that I know of

I think this is what I was looking for: https://github.com/ankidroid/Anki-Android/wiki/AnkiDroid-Javascript-API#stop-speech

mikehardy commented:
Ah yes! That will stop (and you can restart) but unfortunately it stops the whole thing as a unit and will restart from the beginning. There's no way to pause say 4 words in to a 8 word phrase then resume on the 5th word etc. Either way, glad you found something that does handle your use case :-)

mnhende2 commented:
Super excited this is going to be in the next update!  Thanks to everyone who has worked on this over multiple years!  And wishing you all the best in getting the next update finalized!

ODAWGG commented:
Hi,

Firstly thank you to all Anki contributors, Anki has transformed my life, so thank you ! 

Just wondering if this feature been implemented ? I was using it on Android when it first came out and it was great ! I use Anki for music practise so its pefect for pausing long passages. However now after the update I can no longer pause audio and I see it listed a bug.  

Is there something I am missing.

snowtimeglass commented:
This feature was implemented once, but it has been changed because of inconsistency behavior with Anki Desktop.

#14221 
#14296

ODAWGG commented:
Ah, understood. I will stick to this older build then as pausing audio is such a key feature for me. I  wonder how long I can hold out before I have to update ?
